pwiz_tools/Skyline/Model/TransitionDocNode.cs

```diff
@@ -555,9 +556,90 @@ public SkylineDocumentProto.Types.Transition ToTransitionProto(SrmSettings setti
                 }
                 transitionProto.LinkedIons.Add(linkedIon);
             }
+
+            double? ce = GetCollisionEnergy(settings, nodePep, nodeGroup);
+            double? dp = GetDeclusteringPotential(settings, nodePep, nodeGroup);
+
+            if (ce.HasValue)
+            {
+                transitionProto.CollisionEnergy = DataValues.ToOptional(ce);
+            }
+
+            if (dp.HasValue)
+            {
+                transitionProto.DeclusteringPotential = DataValues.ToOptional(dp);
+            }
+
             return transitionProto;
         }
 
+        public double? GetCollisionEnergy(SrmSettings settings, PeptideDocNode nodePep, TransitionGroupDocNode nodeGroup)
+        {
+            double? ce = null;
+            TransitionPrediction predict = settings.TransitionSettings.Prediction;
+            var optimizationMethod = predict.OptimizedMethodType;
+            var lib = predict.OptimizedLibrary;
+            if (lib != null && !lib.IsNone)
+            {
+                var optimization = lib.GetOptimization(OptimizationType.collision_energy,
+                    settings.GetSourceTarget(nodePep), nodeGroup.PrecursorAdduct,
+                    FragmentIonName, Transition.Adduct);
+                if (optimization != null)
+                {
+                    ce = optimization.Value;
+                }
+            }
+
+            double regressionMz = settings.GetRegressionMz(nodePep, nodeGroup);
+            var ceRegression = predict.CollisionEnergy;
+            if (optimizationMethod == OptimizedMethodType.None)
+            {
+                if (ceRegression != null && !ce.HasValue)
+                {
+                    ce = ceRegression.GetCollisionEnergy(nodeGroup.PrecursorAdduct, regressionMz);
+                }
+            }
+            else
+            {
+                if (!ce.HasValue)
+                {
+                    ce = OptimizationStep<CollisionEnergyRegression>.FindOptimizedValue(settings,
+                        nodePep, nodeGroup, this, optimizationMethod, ceRegression,
+                        SrmDocument.GetCollisionEnergy);
+                }
+            }
+
+            if (ExplicitValues.CollisionEnergy.HasValue)

```

Comment:
Usually this would go at the beginning of the function, and the function would return the explicit value early, instead of going through the other code paths of calculating it.
(probably doesn't make any difference, so this way is fine too)

Reply:
Thanks.  I moved it to the beginning.